tla-io/src/test/scala/at/forsyte/apalache/io/itf/TestItfToTla.scala

```diff
@@ -0,0 +1,480 @@
+import at.forsyte.apalache.tla.lir._
+import at.forsyte.apalache.tla.types.tla
+import org.junit.runner.RunWith
+import org.scalatest.funsuite.AnyFunSuite
+import org.scalatestplus.junit.JUnitRunner
+
+@RunWith(classOf[JUnitRunner])
+class TestItfToTla extends AnyFunSuite {
+
+  val itfToTla = new ItfToTla(UJsonScalaFactory)
+
+  import ujson._
+
+  test("validateShapeAndGetTypes") {
+
+    val empty = UJsonRep(Obj())
+    assert {
+      itfToTla.validateShapeAndGetTypes(empty).isLeft
+    }
+
+    val metaEmpty = UJsonRep(Obj(ItfToTla.META_FIELD -> Obj()))
+    assert {
+      itfToTla.validateShapeAndGetTypes(metaEmpty).isLeft
+    }
+
+    val typesNotObj = UJsonRep(
+        Obj(
+            ItfToTla.META_FIELD ->
+              Obj(
+                  ItfToTla.VAR_TYPES_FIELD -> 42
+              ),
+            ItfToTla.VARS_FIELD -> Arr(),
+        )
+    )
+
+    assert {
+      itfToTla.validateShapeAndGetTypes(typesNotObj).isLeft
+    }
+
+    val noVars = UJsonRep(
+        Obj(
+            ItfToTla.META_FIELD ->
+              Obj(
+                  ItfToTla.VAR_TYPES_FIELD ->
+                    Obj(
+                        "x" -> "Int"
+                    )
+              ),
+            ItfToTla.VARS_FIELD ->
+              Arr(), // empty
+        )
+    )
+
+    assert {
+      itfToTla.validateShapeAndGetTypes(noVars).isLeft
+    }
+
+    val noTypes = UJsonRep(
+        Obj(
+            ItfToTla.META_FIELD ->
+              Obj(
+                  ItfToTla.VAR_TYPES_FIELD -> Obj() // empty
+              ),
+            ItfToTla.VARS_FIELD -> Arr("x"),
+        )
+    )
+
+    assert {
+      itfToTla.validateShapeAndGetTypes(noTypes).isLeft
+    }
+
+    val correct = UJsonRep(
+        Obj(
+            ItfToTla.META_FIELD ->
+              Obj(
+                  ItfToTla.VAR_TYPES_FIELD ->
+                    Obj(
+                        "x" -> "Int",
+                        "y" -> "Str -> Bool",
+                    )
+              ),
+            ItfToTla.VARS_FIELD -> Arr("x", "y"),
+        )
+    )
+
+    assert {
+      itfToTla
+        .validateShapeAndGetTypes(correct)
+        .contains(
+            Map(
+                "x" -> IntT1,
+                "y" -> FunT1(StrT1, BoolT1),
+            )
+        )
+    }
+
+  }
+
+  test("attemptUnserializable") {
+
+    val notUS = UJsonRep(Obj())
+
+    assert(itfToTla.attemptUnserializable(notUS).isEmpty)
+
+    def singleUS(v: String): UJsonRep = UJsonRep(
+        Obj(
+            ItfToTla.UNSERIALIZABLE_FIELD -> v
+        )
+    )
+
+    val bogusUS = singleUS("") // illegal identifier
+
+    assert {
+      itfToTla.attemptUnserializable(bogusUS).exists(_.isLeft)
+    }
+
+    val int = singleUS("Int")
+
+    assert {
+      itfToTla.attemptUnserializable(int).exists(_.isLeft)
+    }
+
+    val nat = singleUS("Nat")
+
+    assert {
+      itfToTla.attemptUnserializable(nat).exists(_.isLeft)
+    }
+  }
+
+  test("typeDrivenBuild") {
```

Comment:
This test case is really long. We get much better feedback from tests if we can break them into smaller units, so that when we get "test case 'such and such property should hold' failed" we know from the name of the test what has failed.

It's not essential, but imo these tests will be much more helpful in case of breakage if they are broken into units that note what they are testing. Seems like the natural categorization based on your current tests is to break each type into its own test case?

This basically just amounts to replacing the (tbh, slightly cryptic ;) ) comments with test descriptions.

Reply:
Sure, I can make tighter groups, np.